File: Python/ByTier/B/1965_B_Missing_Subsequence_Sum.py
# Problem: CF 1965 B - Missing Subsequence Sum
# https://codeforces.com/contest/1965/problem/B

"""
Missing Subsequence Sum Problem

Algorithm/Technique:
- The approach uses binary representation to construct a sequence that avoids a specific sum k
- It identifies a bit position where 2^i < k <= 2^(i+1) and uses this to construct values
- The idea is to use powers of 2 for most elements and tweak one element to avoid forming sum k
- This ensures all values from 1 to n except k can be formed as subsequence sums

Time Complexity: O(1) per test case - only a fixed number of operations
Space Complexity: O(1) - only storing a fixed-size array of at most 25 elements

The constructed sequence ensures:
1. No subsequence sums to k
2. All other values from 1 to n (excluding k) can be formed as subsequence sums
"""

t = int(input())

for tc in range(t):
    n, k = map(int, input().split())

    # Find the largest power of 2 such that 2^i < k
    i = 0
    while (1 << (i + 1)) <= k:
        i = i + 1

    # Construct the answer array
    # First element ensures we can form values around k
    # Then add k+1 and k+1 + 2^i to cover range
    # Then add powers of 2 (except the one at index i)
    ans = [k - (1 << i), k + 1, k + 1 + (1 << i)]

    # Add all powers of 2 from 0 to 19 except the one at position i
    for j in range(20):
        if j != i:
            ans.append(1 << j)

    print(len(ans))
    print(*ans)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1965/1965_B_Missing_Subsequence_Sum.cpp
// Problem: CF 1965 B - Missing Subsequence Sum
// https://codeforces.com/contest/1965/problem/B

/*
Algorithm/Technique: Constructive Greedy Approach
Purpose: To find a sequence of at most 25 non-negative integers such that:
- There is no subsequence with sum equal to k.
- For every value v from 1 to n (excluding k), there exists a subsequence with sum equal to v.

Time Complexity: O(log n) per test case, due to bit manipulation and loop over powers of 2.
Space Complexity: O(1) extra space excluding output storage, as the maximum size of the vector is bounded by 25.
*/

#include <stdio.h>
#include <algorithm>
#include <vector>

using namespace std;
int T, n, k;
int main() {
  scanf("%d", &T);
  while (T--) {
    vector<int> v; // Vector to store the constructed sequence
    scanf("%d%d", &n, &k);
    
    int x = k - 1; // We want to represent all numbers from 1 to k-1 using bits
    
    // Fill the vector with powers of 2 up to x, subtracting them from x
    for (int i = 0; 1 << i <= x; i++)
      x -= 1 << i, v.push_back(1 << i);
    
    // If x is still greater than 0, it means we have leftover bits to represent
    if (x)
      v.push_back(x);
      
    // Add k+2 to ensure no subsequence sums to k (since we won't be able to form k now)
    v.push_back(k + 2);
    
    // Add powers of 2 starting from k+1 up to n to cover all values up to n
    for (int i = k + 1; i <= n; i *= 2)
      v.push_back(i);
      
    printf("%d\n", v.size()); // Print length of the sequence
    for (int x : v)
      printf("%d ", x); // Print each element
    puts(""); // Newline after printing the sequence
  }
}


// https://github.com/VaHiX/CodeForces/